File: Python/1787_MaketheXORofAllSegmentsEqualtoZero.py
"""
You are given an array nums​​​ and an integer k​​​​​. The XOR of a segment [left, right] where left <= right is the XOR of all the elements with indices between left and right, inclusive: nums[left] XOR nums[left+1] XOR ... XOR nums[right].

Return the minimum number of elements to change in the array such that the XOR of all segments of size k​​​​​​ is equal to zero.


Example 1:

Input: nums = [1,2,0,3,0], k = 1
Output: 3
Explanation: Modify the array from [1,2,0,3,0] to from [0,0,0,0,0].
Example 2:

Input: nums = [3,4,5,2,1,7,3,4,7], k = 3
Output: 3
Explanation: Modify the array from [3,4,5,2,1,7,3,4,7] to [3,4,7,3,4,7,3,4,7].
Example 3:

Input: nums = [1,2,4,1,2,5,1,2,6], k = 3
Output: 3
Explanation: Modify the array from [1,2,4,1,2,5,1,2,6] to [1,2,3,1,2,3,1,2,3].
 

Constraints:

1 <= k <= nums.length <= 2000
​​​​​​0 <= nums[i] < 210
"""


from typing import List
from collections import Counter
class Solution:
    def minChanges(self, nums: List[int], k: int) -> int:
        """
        Thoughts: 
        there are len(nums) - k + 1 groups.
        we can convert the list with at most len(nums) - k + 1 modifications, at least 0 modifications
        """
        len_nums = len(nums)
        if k == 1:
            return sum([_n > 0 for _n in nums])
        kmers = Counter()
        for i in range(len_nums - k + 2):
            kmers[tuple(sorted(nums[i: i + k - 1]))] += 1
        print(kmers)
        m_keys = kmers.most_common()[0]
        val = 0
        for _k in m_keys:
            val ^= _k
        for i in range(len_nums - k + 2):
            pass



# 作者：yuan-zhi-b
# 链接：https://leetcode-cn.com/problems/make-the-xor-of-all-segments-equal-to-zero/solution/python3-dong-tai-gui-hua-yi-huo-jie-guo-6kkyf/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。
class Solution:
    def minChanges(self, nums: List[int], k: int) -> int:
        freq=defaultdict(Counter)
        for i in range(k):
            for j in range(i,len(nums),k):
                freq[i][nums[j]]+=1
        dp=[[math.inf]*(2**11) for _ in range(len(freq)+1)]
        dp[-1][0]=0
        for i in range(k-1,-1,-1):
            mindp=min(dp[i+1])
            itemCnt=sum(freq[i].values())
            for xor in range(1025):
                res=itemCnt+mindp
                for key,cnt in freq[i].items(): # change all nums in 'i' set to v
                    res=min(res,itemCnt-cnt+dp[i+1][xor^key])
                dp[i][xor]=res
        return dp[0][0]



from typing import List
from collections import Counter
class Solution:
    def minChanges(self, nums: List[int], k: int) -> int:        

S = Solution()
nums = [1,2,0,3,0]
k = 1
print(S.minChanges(nums, k))
nums = [3,4,5,2,1,7,3,4,7]
k = 3
print(S.minChanges(nums, k))
nums = [1,2,4,1,2,5,1,2,6]
k = 3
print(S.minChanges(nums, k))

File: CPP/1787_MaketheXORofAllSegmentsEqualtoZero.cpp
const int INF = 0x3f3f3f3f;

class Solution {
public:
    int minChanges(vector<int>& nums, int k) {
        int n = nums.size();
        vector<unordered_map<int, int>> groups(k);
        vector<int> size(k);
        for (int i = 0; i < k; ++i) {
            for (int j = i; j < n; j += k) {
                groups[i][nums[j]]++;
                size[i]++;
            }
        }
        //异或值为random，需要操作的最小次数
        vector<int> dp(1 << 10, INF);
        dp[0] = 0;
        //从左到右，考虑K区间的index i的处理
        //即在nums中所有映射在区间k中index 为 i的集合的元素修改方法
        for (int i = 0; i < k; ++i) {
            //讨论第一种情况，即i对应集合的值都修改为一个不存在该集合中的值，那么这个集合就要修改size[i]次
            /**
            dp的涵义：在当前index前，异或一个值需要操作的最小次数
            eg: 
            1. 此时i = 3, 在操作前三个集合，即[0, 1, 2]后，会获取多个异或值，并且都是此时获取该异或值的最小操作次数
            2. 要想修改i=3的集合后，操作最小，就在之前获取异或值中最小操作次数的基础上 + size[i];
                比如dp中此时获得55需要的次数最小，那么之后就以这个为基础，在 i=3对应的集合上修改
                修改后可以为任意值， 比如为获取99，只需将i=3的区间所有值都修改为 99^55就好
            */
            int lo = *min_element(dp.begin(), dp.end());
            vector<int> ndp(1 << 10, lo + size[i]);


            // 讨论第二种情况
            /**
            1. 之前考虑的都是集合i中元素都修改为集合i中没有的数字，那么对于集合i而言，它的修改次数是固定的，
                    所以只要求之前需要修改的次数的最小值就可以获取最小值
            2. 可是存在另一种情况，即修改值存在集合i中，那么对于集合i修改的次数就是变量，为 size[i] - group[i][val]
                1. 在修改集合i的次数值为变值时，就需要枚举全部情况
                2. 以val^val`为例
                    1. 因为可能存在 之前集合操作变成 val`的次数不是最小，但是加上size[i] - group[i][val]后
                        其值比 size[i] + min(dp[random])要小
            */
            for (int j = 0; j < (1 << 10); ++j) {
                if (dp[j] == INF)
                    continue;
                for (auto [p, freq] : groups[i]) {
                    int nxt = p ^ j;
                    ndp[nxt] = min(ndp[nxt], dp[j] + size[i] - freq);
                }
            }
            // 将ndp copy给dp，表示操作完i集合后，异或得到每个可能的值需要的最小操作数
            dp = move(ndp);
        }
        //从集合0开始到集合k-1，所有集合都讨论处理后，dp[0]就是获取0需要操作的最小次数
        return dp[0];
    }
};